src/ucp/core/ucp_am.c

```diff
@@ -508,69 +572,91 @@ ucp_am_send_req(ucp_request_t *req, size_t count,
     if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
-        ucs_trace_req("releasing send request %p, returning status %s", req,
-                      ucs_status_string(status));
-        ucp_request_put(req);
-        return UCS_STATUS_PTR(status);
+        ucp_request_imm_cmpl_param(param, req, status, send);
     }
 
-    ucp_request_set_callback(req, send.cb, (ucp_send_nbx_callback_t)cb, NULL);
+    ucp_request_set_send_callback_param(param, req, send);
 
     return req + 1;
 }
 
-UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nb,
-                 (ep, id, payload, count, datatype, cb, flags),
-                 ucp_ep_h ep, uint16_t id, const void *payload,
-                 size_t count, ucp_datatype_t datatype,
-                 ucp_send_callback_t cb, unsigned flags)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_send_inline(ucp_ep_h ep, uint16_t id, uint32_t flags, const void *header,
+                   size_t header_length, const void *buffer, size_t length)
+{
+    if (ucs_likely(!(flags & UCP_AM_SEND_REPLY) &&
+                   ((length == 0) || (header_length == 0)) &&
+                   ((ssize_t)(length + header_length) <=
+                    ucp_ep_config(ep)->am.max_short))) {
+        return ucp_am_send_short(ep, id, flags, header, header_length,
+                                 buffer, length);
+    }
+
+    return UCS_ERR_NO_RESOURCE;
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_common,
+                 (ep, id, header, header_length, buffer, count, param, priv_flags),
+                 ucp_ep_h ep, unsigned id, const void *header,
+                 size_t header_length, const void *buffer, size_t count,
+                 const ucp_request_param_t *param, unsigned priv_flags)
 {
     ucs_status_t status;
     ucs_status_ptr_t ret;
+    ucp_datatype_t datatype;
     ucp_request_t *req;
-    size_t length;
+    uint32_t attr_mask;
+    uint32_t flags;
 
     UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_AM,
                                     return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
 
-    if (ucs_unlikely((flags != 0) && !(flags & UCP_AM_SEND_REPLY))) {
-        return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
-    }
-
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
 
-    if (ucs_likely(!(flags & UCP_AM_SEND_REPLY)) &&
-        (ucs_likely(UCP_DT_IS_CONTIG(datatype)))) {
-        length = ucp_contig_dt_length(datatype, count);
-
-        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
-            status = ucp_am_send_short(ep, id, payload, length);
-            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-                UCP_EP_STAT_TAG_OP(ep, EAGER);
-                ret = UCS_STATUS_PTR(status);
-                goto out;
-            }
+    flags     = priv_flags | ucp_request_param_flags(param);
+    attr_mask = param->op_attr_mask &
+                (UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FLAG_NO_IMM_CMPL);
+
```

Comment:
remove space line

Reply:
why? Typically we set new line before `if`, like in `ucp_tag_send_nbx`